libs/server/AOF/AofProcessor.cs

```diff
@@ -265,22 +310,70 @@ private unsafe bool ReplayOp(byte* entryPtr)
                     ObjectStoreDelete(objectStoreBasicContext, entryPtr);
                     break;
                 case AofEntryType.StoredProcedure:
-                    RunStoredProc(header.type, customProcInput, entryPtr);
+                    RunStoredProc(type, customProcInput, entryPtr, isHistoricMode: false);
                     break;
                 default:
-                    throw new GarnetException($"Unknown AOF header operation type {header.opType}");
+                    throw new GarnetException($"Unknown AOF header operation type {opType}");
             }
             return true;
         }
 
-        unsafe void RunStoredProc(byte id, CustomProcedureInput customProcInput, byte* ptr)
+        private unsafe bool ReplayLegacyOp(AofEntryType opType, long aofHeaderRecordVersion, byte type, byte* entryPtr)
         {
-            var curr = ptr + sizeof(AofHeader);
+            // Skips versions that were part of checkpoint
+            if (SkipRecord(opType, aofHeaderRecordVersion)) return false;
 
-            // Reconstructing CustomProcedureInput
+            switch (opType)
+            {
+                case AofEntryType.StoreUpsert:
+                    LegacyUpsert(entryPtr, storeInput);
+                    break;
+                case AofEntryType.StoredProcedure:
+                    RunStoredProc(type, customProcInput, entryPtr, isHistoricMode: true);
+                    break;
+                default:
+                    throw new GarnetException($"Unknown AOF header operation type {opType} for AOF legacy processing");
+            }
 
-            // input
-            customProcInput.DeserializeFrom(curr);
+            return true;
+        }
+
+        private unsafe void LegacyUpsert(byte* ptr, RawStringInput storeInput)
+        {
+            ref SpanByte key = ref Unsafe.AsRef<SpanByte>(ptr + sizeof(LegacyAofHeader));
+            // input is willingly being ignored from older format
+            ref SpanByte input = ref Unsafe.AsRef<SpanByte>(ptr + sizeof(LegacyAofHeader) + key.TotalSize);
+            ref SpanByte value = ref Unsafe.AsRef<SpanByte>(ptr + sizeof(LegacyAofHeader) + key.TotalSize + input.TotalSize);
+
+            /* Reconstructing RawStringInput based on the following invariant:
+               - extra metadata on value tells us there was an expiration
+            */
+            long expiration = value.ExtraMetadata;
+            if (value.MetadataSize == 0)
+            {
+                storeInput = new RawStringInput(RespCommand.SETEX, 0);
+            }
+            else
+            {
+                storeInput = new RawStringInput(RespCommand.SETEX, 0, expiration);
+            }
+
+            StoreUpsertPostParsing(basicContext, ref key, storeInput, ref value);
+        }
+
+        unsafe void RunStoredProc(byte id, CustomProcedureInput customProcInput, byte* ptr, bool isHistoricMode)
```

Comment:
Might want to stick to the "isLegacyFormat" naming

Reply:
agreed